metadata/metadata.go

```diff
@@ -159,20 +171,38 @@ func AppendToOutgoingContext(ctx context.Context, kv ...string) context.Context
 	return context.WithValue(ctx, mdOutgoingKey{}, rawMD{md: md.md, added: added})
 }
 
-// FromIncomingContext returns the incoming metadata in ctx if it exists.  The
+// FromIncomingContext returns the incoming metadata in ctx if it exists. The
 // returned MD should not be modified. Writing to it may cause races.
 // Modification should be made to copies of the returned MD.
-func FromIncomingContext(ctx context.Context) (md MD, ok bool) {
-	md, ok = ctx.Value(mdIncomingKey{}).(MD)
-	return
+//
+// All keys in the return MD are lowercase.
+func FromIncomingContext(ctx context.Context) (MD, bool) {
+	md, ok := ctx.Value(mdIncomingKey{}).(MD)
+	if !ok {
+		return nil, false
+	}
+	out := MD{}
+	for k, v := range md {
+		// We need to manually convert all keys to lower case, because MD is a
+		// map, and there's no guarantee that the MD attached to the context is
+		// created using our helper functions.
+		key := strings.ToLower(k)
+		out[key] = v
+	}
```

Comment:
I think I'm okay with this, even though it's a performance hit (new copy & lowercase).  Headers shouldn't be hyper performance sensitive, and we can remove the clause about how "The returned MD should not be modified", since we now return a copy.

Reply:
> remove the clause about how "The returned MD should not be modified"

Removing means we will almost never be able to add it back. Hmm, would we regret later?

Also, I just noticed it's the same for `FromOutgoingContext`, and we have had it for a while.

Reply:
If we specify that all returned keys are lowercase, then we will always need to copy, won't we?

Reply:
Doc updated

Reply:
hey, so this created a change in behavior, now that we can't modify the incoming context anymore, which we were doing to put additional values in the context. What's the supported way of modifying the incoming context? or is it supported at all?

Reply:
If you are talking about the incoming metadata in the context, you can reattached the context with additional value using `NewIncomingContext()`.